### ISSUE TITLE
"There are too many connected processes" in using cljs-repl

### ISSUE THREAD

**Issue by chenyong** [thheller/shadow-cljs#185]
I got watch-compiling and Node.js process running then started using a REPL to connect it:

```
=>> yarn shadow-cljs cljs-repl app
yarn run v1.3.2
$ /Users/chen/repo/mvc-works/termina/server/node_modules/.bin/shadow-cljs cljs-repl app
shadow-cljs - config: /Users/chen/repo/mvc-works/termina/server/shadow-cljs.edn version: 2.0.144
shadow-cljs - connected to server
[2:1]~cljs.user=> (.log js/console 'conected')
There are too many connected processes.
```

Long time not trying to use REPL, but found myself need to look into states... Is it the right way to connect to the Node REPL?

[config file](https://github.com/mvc-works/termina/blob/master/server/shadow-cljs.edn):

```edn
{:source-paths ["src"],
 :http {:port 9640}
 :dependencies [[mvc-works/hsl    "0.1.2"]
                [cumulo/recollect "0.4.3"]]
 :builds {:app {:target :node-script
                :output-to "target/main.js"
                :main server.main/main!
                :devtools {:after-load server.main/reload!}}}}
```

**Comment by thheller** [comment]
> There are too many connected processes.

Means there are more than one node processes running your code.

Suppose you do:

```
$ shadow-cljs watch app
# Client #1, run node app
$ shadow-cljs cljs-repl app
cljs.user=> (def x 1)

# Client #2, run second node app
cljs.user=> (inc x)
```

There are now 2 clients where the REPL could eval code. Both have loaded the compiled JS but they are in different states. Client 2 does not have `x` since it wasn't around when REPL eval'd it. So it cannot eval `(inc x)`. The compiler however only has one compiler state so you will not get a warning since it things `x` exists.

One solution is to ignore this problem and always eval everything typed into the REPL in all connected Clients (figwheel does this). Which result do you display? Client 1 evals fine, Client 2 throws an error. It is undefined which client will answer first. Do you only send to the first connection? The last?

I opted for safety and just warn if multiples clients are connected. Ideally you would pick which one to send to but how do you pick which one?

The REPL already is in a funky state if you `(def x 1)` and then restart the only node process. The compiler env still knows `x` but the Client doesn't. Same deal in the browser, reloading causes you to lose only half the state.

I can change the default "safe" behavior and broadcast to everything if someone can tell me a reliable way to deal with results. nREPL for example does not deal well with multiple results.

**Comment by chenyong** [comment]
Roughly got the problem of inconsistent states... not very familiar with REPLs though.

Probably I'm using cljs-repl in the wrong way. I want to connect to a running Node process(or a browser running), what's the right steps to do it?

**Comment by thheller** [comment]
- Run ONE client process (node, browser).
- Run ONE `watch` process.
- Run as many `shadow-cljs cljs-repl app` processes as you like.

Done. You have everything setup correctly. There are just 2 client processes connected which currently is not allowed.

**Comment by chenyong** [comment]
Found something strange. I tried to start from refresh and it worked. But I reproduced the error again after that, not sure which step caused.

But seeing from the logs, I found no "disconnection" log when I kill my nodejs process, which I did many times.

```
=>> yarn watch
yarn run v1.3.2
$ shadow-cljs watch app
shadow-cljs - config: /Users/chen/repo/mvc-works/termina/server/shadow-cljs.edn version: 2.0.144
shadow-cljs - starting ...
shadow-cljs - server running at http://0.0.0.0:9640
shadow-cljs - watching build :app
[:app] Configuring build.
[:app] Compiling ...
[:app] Build completed. (63 files, 1 compiled, 0 warnings, 2.90s)
JS runtime connected.
JS runtime connected.
JS runtime connected.
JS runtime connected.
JS runtime connected.
```

Is it possible that I'm actually running only 1 client, but shadow-cljs detects no disconnection and thinks I'm running several of them?

**Comment by thheller** [comment]
Hmm indeed. node processes don't get cleaned up properly. Will look into it.

**Comment by chenyong** [comment]
Do you think this issue related? https://github.com/websockets/ws/issues/1256#issuecomment-360115197

I saw this error in my WebSocket connection days before. I noticed shadow-cljs also uses `ws`.

**Comment by thheller** [comment]
That looks like an issue with using `ws` as a server. We only use the client on `node`.

**Comment by thheller** [comment]
Fixed in `shadow-cljs@2.0.146`.
